doc/proposals/scorecard-custom-tests-2.md

```diff
@@ -0,0 +1,261 @@
+
+Scorecard would support the creation of a Kuttl test image.  This custom test would execute kuttl to perform testing enabling scorecard users a means to write tests using kuttl but have the result integrated into scorecard results.  
+
+## Design Details
+
+### Custom Test Definition
+
+* A custom test is any sort of test logic required, written in any programming language that satifies the scorecard test requirements for output format and execution.
+
+### Custom Test Components
+
+This proposal outlines the end-user facing components that make up a custom test including:
+
+#### Test Image
+
+A custom test is based on a container image, that image is created by
+the end-user using whatever tools they want to use.
+
+The test image must produce container log output that conforms to the scorecard Test Result output.  The output is an array of Test Results, so a test image could contain multiple tests if necessary.  Here is a sample of test output in the required format:
+```
+    [
+      {
+	      "name": "Spec fields with descriptors",
+	      "description": "All spec fields have matching descriptors in the CSV",
+	      "state": "fail",
+	      "suggestions": [
+		"Add a spec descriptor for size"
+	      ],
+	      "crname": "example-memcached"
+      }, 
+      {
+	      "name": "Some other test",
+	      "description": "Another test on the CSV",
+	      "state": "pass",
+	      "suggestions": [
+	      ],
+	      "crname": "example-memcached"
+      },
+    ]
+```
+
+The test image will be executed as a Pod by the scorecard with a restart policy of `never`.
+
+When executing a scorecard test, scorecard will create a ConfigMap holding the scorecard configuration manifests such as the Custom Resource or Bundle being tested.  This ConfigMap can then be mounted into the custom test image and used within the test execution.  
+
+The ConfigMap would be mounted into the test container at the following location:
+```
+/scorecard/config
+```
+
+For the case of testing a bundle image, the following command will run scorecard tests which are configured within the bundle image itself:
+```
+operator-sdk scorecard --bundle-image=quay.io/myorg/mytest
+```
+
+In this case, for each test execution, we could construct:
+
+ * A ConfigMap containing the required test configuration for the test.
+ * A PodSpec with the containers being the bundle image, the test image, and an init container that copies the bundle files from the bundle image into the test image at a specific place (using a shared emptyDir volume mount) (e.g. /scorecard/bundle)
+ * We would also mount the ConfigMap into the test image container (e.g. /scorecard/config)
+
+For this, scorecard would need to support bundle images which is included as a User Story in this proposal.  If the bundle image were to contain the scorecard configuration it might be possble to use that for running the scorecard test as opposed to having a scorecard configuration outside of the bundle image being used.
```

Comment:
I think we should talk through this a little more.

Can the bundle image be the one and only required command line flag that dictates where to find the configuration for what to test? Put another way, are there any other mechanisms that we want (or need) to support to enable scorecard to know what to test (e.g. local filesystem directory). 

IMO, I think we start out assuming that the bundle image will contain all the necessary info and configs to run the full suite.

There are three classes of configuration in my mind:
1. Configuring `operator-sdk scorecard` to know what test images to run.
2. Configuring each test image with the bundle contents.
3. Configuring each test image with its custom configuration.

If the assumption is that the bundle image actually contains the first class of configuration, that means that scorecard will need to be able to fetch and extract the contents of the bundle image locally so that it can know what test images to run in the cluster. This could be a problem in the CI/pipeline case where access to a Docker socket is not available. Are there other OCI tools or libraries available that can pull and extract images without requiring root?

Then we have the question about the method of providing configuration to the test images. It looks like there are a couple of different methods listed here. Do we have enough information now to choose one of those?

**Method 1** Add bundle image container as sidecar, use initContainer to copy bundle files into emptyDir shared with the test container.

This method is desirable because it doesn't require Docker (or any other OCI utilities) to exist locally. This seems like a big benefit for the CI/pipeline use case, but maybe we need those tools locally anyway if we want scorecard to get its test configuration from the bundle itself?

**Method 2** Scorecard pulls and extracts the bundle image locally, creates a ConfigMap of the contents, and launches the test container with the ConfigMap mounted in.

This method is more straightforward on the PodSpec side since it just involves creating a ConfigMap and it in to the test container.

Reply:
I wonder if people will not like having their test assets (scorecard config + test manifests) stored within something like a operator bundle?   Some places I've worked at would not want to ship tests within a production bundle for a variety of reasons.  But if thats not an issue, then Method 1 seems nice because it keeps you from having test assets laying about on local disk or ConfigMaps being created and then having to be cleaned up.

Reply:
I suppose for a production bundle, you could just strip out the scorecard assets as some sort of workflow or CLI command.

Reply:
@shawn-hurley @dmesser Thoughts?

Reply:
I think we should support both scenarios - test configuration shipped in parallel with the bundle and as part of the bundle. The latter is just making everything very portable and simplifies the UX. Like @joelanford said, it should be possible to just give scorecard the bundle and it knows, what to do.
It should be noted that the Operator bundle is not the actual production application itself. Pods running on cluster will not have access to any test scripts or manifests. See [bundle extraction on cluster](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/pull-bundle-on-a-cluster.md) for more info.
There is precedence for shipping test configurations with the application packaging with `helm`.

If some pipeline / organization absolutely cannot deal with test scripts being part of the bundle, there should be a second option to instruct scorecard where to find the data. The tradeoff for them is that it makes those people jobs harder to orchestrate a test run and find the right version of the test configuration for the bundle at hand.

Reply:
Seems reasonable. How about this:

* Add a `--test-asset-image` flag.
* The test asset image must have the exact same filesystem layout for the scorecard tests as the bundle image.
* If `--test-asset-image` is provided, scorecard get all configs from that image and none from the bundle image. This covers the case that the bundle image ALSO has test assets in it.